File: problems/290-word-pattern/solution.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False
        p_map = {}
        w_map = {}
        for i in range(len(pattern)):
            p = pattern[i]
            w = words[i]
            if p not in p_map:
                if w in w_map:
                    return False
                p_map[p] = w
                w_map[w] = p
            elif p_map[p] != w:
                return False
        return True

File: problems/290-word-pattern/solution.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<char, string> p2s;
        unordered_map<string, char> s2p;
        stringstream ss(s);
        string word;
        int i = 0;
        while (ss >> word) {
            if (i >= pattern.size()) return false;
            char c = pattern[i++];
            if (p2s.count(c) && p2s[c] != word) return false;
            if (s2p.count(word) && s2p[word] != c) return false;
            p2s[c] = word;
            s2p[word] = c;
        }
        return i == pattern.size();
    }
};